File: 2327-number-of-people-aware-of-a-secret/2327-number-of-people-aware-of-a-secret.py
class Solution:
    def peopleAwareOfSecret(self, n: int, delay: int, forget: int) -> int:
        know, share = deque([(1, 1)]), deque()
        know_cnt, share_cnt = 1, 0

        for i in range(2, n + 1):
            if know and know[0][0] == i - delay:
                know_cnt -= know[0][1]
                share_cnt += know[0][1]
                share.append(know[0])
                know.popleft()
            
            if share and share[0][0] == i - forget:
                share_cnt -= share[0][1]
                share.popleft()
            
            if share:
                know_cnt += share_cnt
                know.append((i, share_cnt))

        return (know_cnt + share_cnt) % (10 ** 9 + 7)

File: 2327-number-of-people-aware-of-a-secret/2327-number-of-people-aware-of-a-secret.cpp
class Solution {
private:
    static constexpr int mod = 1'000'000'007;

public:
    int peopleAwareOfSecret(int n, int delay, int forget) {
        int knowCnt = 1, shareCnt = 0;
        queue<pair<int, int>> know, share;
        know.push({1, 1});

        for (int i = 2; i <= n; ++i){
            if (know.front().first == i - delay){
                knowCnt = (knowCnt - know.front().second + mod) % mod;
                shareCnt = (shareCnt + know.front().second) % mod;
                share.push(know.front());
                know.pop();
            }

            if (share.front().first == i - forget){
                shareCnt = (shareCnt - share.front().second + mod) % mod;
                share.pop();
            }

            if (!share.empty()){
                knowCnt = (knowCnt + shareCnt ) % mod;
                know.push({i, shareCnt});
            }
        }

        return (knowCnt + shareCnt) % 1'000'000'007;
    }
};